_adr/88/index.adoc

```diff
@@ -1,48 +1,70 @@
+  - "sso"
 applies_padrs: # What PADRs does this ADR apply?
 applies_patterns: # What APs does this ADR apply?
 ---
 
-// Top style tips:
-// * Use one sentence per line
-// * No unexpanded acronyms
-// * No undefined jargon
-
-// No need for a title heading, it's added by the template
+:rover-groups: https://rover.redhat.com/groups
 
 ## Context and Problem Statement
-// What is the background against which this decision is being taken?
+The purpose of this ADR is to define a standardized approach for securing admin API in a fleet manager
 
 ## Goals
-// Bulleted list of outcomes that this ADR, if accepted, should help achieve
+* Provide standardized way to secure admin API endpoints for fleet managers
 
 ## Non-goals
-// Bulleted list of outcomes that this ADR is not trying to achieve.
 
 ## Current situation
-// Where are we now?
 
 ## Proposal
-// What is the decision being proposed
+
+image::adr-88-admin-api-architecture-diagram.png[architecture diagram]
+
+A Keycloak realm should be used to authorize access to admin API endpoint(s). 
+Such realm should contain roles that allow access to endpoints to the privileged API users.
+Recommended approach would be to have set of roles per environment and name them in the way that indicates the permissions they grant as follows:
+
+* <unique-fleet-manager-prefix>-admin-read-stage
+* <unique-fleet-manager-prefix>-admin-write-stage
+* <unique-fleet-manager-prefix>-admin-full-stage
+
+* <unique-fleet-manager-prefix>-admin-read-prod
+* <unique-fleet-manager-prefix>-admin-write-prod
+* <unique-fleet-manager-prefix>-admin-full-prod
+
+These roles map to the following permissions for the environment included as the role postfix (e.g. `prod`):
+
+* \*-read-* - list permissions
+* \*-write-* - list/ update permissions
+* \*-full-* - full permissions (list/ update/ delete)
+
+Once Keycloak instance is set up, a privileged user uses their credentials to get a JWT token from Keycloak.
+The JWT token will be sent to a admin API in a HTTP request.
+Admin API authorization middleware verifies token signature with Public JWKS Certificate and roles included in the JWT token are checked against roles specified in the fleet manager.
+Matching role for given admin endpoint will grant access to the action associated with that endpoint.
+Failure to match the required role will result in restricting access to the endpoint.
+
+image::adr-88-authentication-flow.png[architecture diagram]
+
+## Suggested implementation for Red Hat Internal Keycloak
+
+Instead of using dedicated Keycloak instance/ realm, roles allowing to access admin API endpoints are represented by membership in a {rover-groups}[Rover group(s)]
 
 ### Threat model
-// Provide a link to the relevant threat model. 
-// You must either update an existing threat model(s) to cover the changes made by this ADR, or add a new threat model.
+For both, general and Red Hat specific approaches, there will be very limited amount of privileged users.
```

Comment:
you could add that it relies on the existing security model we use for general kafka instance access, so not a new paradygm? I'm assuming what I say is true.

Reply:
well. this is a generic ADR, so I was hesitant to add anything here. would you agree?

Reply:
I understand that you see the threat model as attached to how red hat deploys the stack, correct?

That is a good point and is fair but some parts are inherently part of the architecture of the upstream. I think what I described is in this category (unless I’m over my head which could very well be).

Reply:
I have rewritten it a bit. does it make sense now?

Reply:
yes thank you